### PR TITLE
Adding chaotic systems

### PR DESCRIPTION
**Pull request by MartinuzziFrancesco**
I'm implementing some models from: https://github.com/williamgilpin/dysts 

The description for the documentation at the moment follows the one given in the [.json](https://github.com/williamgilpin/dysts/blob/master/dysts/data/chaotic_attractors.json) file.

The definition of the Thomas attractor seems to be a little different from the one already in the package so I included that also with different names. I will be adding more so this pr should stay open until all the systems are implemented

### CODE REVIEW

**Comment by Datseris** [comment]
Thanks, but here you are implementing **dynamical systems** yet in their documentation strings you are calling them **attractors** instead. Attractors and dynamical systems are different things.

**Comment by MartinuzziFrancesco** [comment]
Yes I'm aware of that. The docs strings at the moment are just the description provided in the original repo and not as detailed as the ones available in the DynamicalSystems docs so I need to revisit them in any case

**Comment by Datseris** [comment]
please use the second way https://github.com/JuliaDynamics/DynamicalSystemsBase.jl/blob/main/src/predefined/continuous_famous_systems.jl#L759  . The old way needs to be updated, slowly but painfully.

**Comment by Datseris** [comment]
and please use multiline if the line exceeds 92 characters, like so 
```
[^Sprott2020]: 
    Sprott, J.C. 'Do We Need More Chaos Examples?', 
    Chaos Theory and Applications 2(2),1-3, 2020
```

**Comment by MartinuzziFrancesco** [comment]
I fixed the citation style for all of the continuous systems examples, and fixed some of the docs for the 92 characters limit. Since it will take me a while to add all the systems from the paper I think that we could review this pr as it stands and merge it if it meets the criteria. I will add some 10 or something systems for each pr in order not to have a large amount of code to review.

Are there tests to be added for the dynamical system examples?

**Comment by Datseris** [comment]
> Are there tests to be added for the dynamical system examples?

No, I have decided to not test this. This is okay, because it is not formally part of the API. In the docs I say that all these systems are provided purely as convenience.

File: src/predefined/continuous_famous_systems.jl [hunk 1]
```diff
@@ -1405,3 +1498,297 @@ end
     du2 = 1/τ*(-N + G)
     return SVector{2}(du1, du2)
 end
+
+"""
+```julia
+sakarya(u0= [-2.8976045, 3.8877978, 3.07465];
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1504]: where does the name `sakarya` come form?

File: src/predefined/continuous_famous_systems.jl [hunk 2]
```diff
@@ -1405,3 +1498,297 @@ end
...
+```
+```math
+\\begin{aligned}
+\\dot{x} &= ax + hy + syz\\\\
+\\dot{y} &= -by - px + qxz \\\\
+\\dot{z} &= cz - rxy
+\\end{aligned}
+```
+A system presenting robust chaos that varies from single wing to double wings to four wings.
+Its attractor arises due to merging of two disjoint bistable attractors [^Li2015].
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1523]: Do you remember at which parameter value thius happens\/ and which is the parameter that is actually varied for this to happen?

Also, is it really necessary that wre keep so many free parameters?

File: src/predefined/continuous_famous_systems.jl [hunk 3]
```diff
@@ -1405,3 +1498,297 @@ end
...
+    h = 1.0,
+    p = 1.0,
+    q = 0.4,
+    r = 0.3,
+    s = 1.0
+)
+    return CDS(loop_sakarya, u0, [a,b,c,h,p,q,r,s])
+end
+
+function loop_sakarya(u, p, t)
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1542]: `loop_sakarya` doesn't make seense. `loop` meant `lorenz out of place`. You should just use `sakarya_rule` as the name of the function.

File: src/predefined/continuous_famous_systems.jl [hunk 4]
```diff
@@ -1405,3 +1498,297 @@ end
...
+\\begin{aligned}
+\\dot{X} &= \\sigma(Y-X)f(X,Y,Z) \\\\
+\\dot{Y} &= (-XZ + \\rho X -Y)f(X,Y,Z) \\\\
+\\dot{Z} &= (XY - \\beta Z)f(X,Y,Z)
+\\end{aligned}
+```
+
+Lorenz system bounded by a confining potential [^SprottXiong2015].
+
+[^SprottXiong2015]: Sprott & Xiong (2015). Chaos.
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1572]: What's the reference here? Just "Chaos" is not enough to find the reference.

File: src/predefined/continuous_famous_systems.jl [hunk 5]
```diff
@@ -1405,3 +1498,297 @@ end
...
+\\dot{z}_2 &= x_2 y_2 - \\beta z_2
+\\end{aligned}
+```
+Two coupled Lorenz attractors [^Lorenz1963].
+
+[^Lorenz1963]:
+    Lorenz, Edward N. Deterministic nonperiodic flow.
+    Journal of the atmospheric sciences 20.2 (1963): 130-141.
+"""
+function lorenz_coupled(u0=[7.0921483,10.117408,25.708428,2.1689893,0.96121877,0.89550187];
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1621]: You can safely remove this system. In DynamicalSystems.jl 3.0 I will create a "CoupledSystem" dynamical system that takes in two systems and couples them either uni or bi-directionally.

File: src/predefined/continuous_famous_systems.jl [hunk 6]
```diff
@@ -1405,3 +1498,297 @@ end
...
+```
+A cyclically-symmetric attractor correspondng to a frictionally-damped
+particle traversing a 3D lattice of forces [^Rene1999].
+
+[^Rene1999]:
+    Thomas, Rene (1999). Deterministic chaos seen in terms of feedback circuits:
+    Analysis, synthesis, labyrinth chaos.
+    Int. J. Bifurc. Chaos.
+"""
+function thomas(u0=[2.199442,2.3634225,3.220197]; a = 1.85, b = 10)
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1665]: We already have this system in the library! `thomas_cyclical`.

File: src/predefined/continuous_famous_systems.jl [hunk 7]
```diff
@@ -1405,3 +1498,297 @@ end
...
+```math
+\\begin{aligned}
+\\dot{r} &= \\frac{p_r}{M+m}\\\\
+\\dot{p}_r &= -Mg + mg\\cos(\\theta)\\\\
+\\dot{\\theta} &= \\frac{p_{\\theta}}{mr^2}\\\\
+\\dot{p}_{\\theta} &= -mgr\\sin(\\theta)
+\\end{aligned}
+```
+A mechanical system consisting of two swinging weights connected by ropes and pulleys.
+This is only chaotic when m2 is sufficiently larger than m1, and there are nonzero initial
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1695]:
```suggestion
This is only chaotic when `m2` is sufficiently larger than `m1`, and there are nonzero initial
```

File: src/predefined/continuous_famous_systems.jl [hunk 8]
```diff
@@ -1405,3 +1498,297 @@ end
...
+
+[^Tufillaro1984]: 
+    Tufillaro, Nicholas B.; Abbott, Tyler A.; Griffiths, David J. (1984).
+    Swinging Atwood's Machine. American Journal of Physics. 52 (10): 895--903.
+"""
+function swinging_atwood(u0=[0.113296,1.5707963267948966,0.10992,0.17747]; m1=1.0, m2=4.5)
+    return CDS(loop_swingingatwood, u0, [m1, m2])
+end
+
+function loop_swingingatwood(u, p, t)
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1706]: same story with `loop`. change it everywhere to `rule`.

File: src/predefined/continuous_famous_systems.jl [hunk 9]
```diff
@@ -1405,3 +1498,297 @@ end
...
+        du1 = a*x - b*y + c*z*x + d*z*(x^2 + y^2)
+        du2 = a*y + b*x + c*z*y
+        du3 = e - z^2 - f*(x^2 + y^2) - a*z^3
+    end
+    return SVector{3}(du1, du2, du3)
+end
+
+"""
+```julia
+halvorsen(u0=[-8.6807408,-2.4741399,0.070775762]; a = 1.4, b = 4.0)
```
Comment (Datseris) [inline src/predefined/continuous_famous_systems.jl:1766]: Where does `halvorsen` come from?

**Review by Datseris — COMMENTED** [review]
Thank you very much, this is a massive help. I agree with you that reviewing a small content is much easier for me!

**Comment by MartinuzziFrancesco** [comment]
> No, I have decided to not test this. This is okay, because it is not formally part of the API. In the docs I say that all these systems are provided purely as convenience.

Wonderful, that does make sense

> Thank you very much, this is a massive help. I agree with you that reviewing a small content is much easier for me!

No problems! I'll address the reviews as soon as possible

File: src/predefined/continuous_famous_systems.jl (hunk 2 above)
Comment (MartinuzziFrancesco) [reply src/predefined/continuous_famous_systems.jl:1523]: No I don't, I'll give a read at the paper again and get back at you.

Regarding the free parameters, I'm following Gilpin's [implementation](https://github.com/williamgilpin/dysts/blob/master/dysts/flows.py#L766) and it seems that not only for this system but also for others (the  [`thomas`](https://github.com/williamgilpin/dysts/blob/master/dysts/flows.py#L83) implementation for example) the amount of free parameters exceeds the systems found in the literature. Should I maintain the amount found in the python implementation or follow the literature equations?

File: src/predefined/continuous_famous_systems.jl (hunk 9 above)
Comment (MartinuzziFrancesco) [reply src/predefined/continuous_famous_systems.jl:1766]: chapter 3.6.1 of Sprott's Elegant Chaos: Algebraically Simple Chaotic Flows

> A simple example of such a circulant system was suggested by Arne Dehli Halvorsen (unpublished)

File: src/predefined/continuous_famous_systems.jl (hunk 2 above)
Comment (Datseris) [reply src/predefined/continuous_famous_systems.jl:1523]: Follow literature equations!

Comment (MartinuzziFrancesco) [reply src/predefined/continuous_famous_systems.jl:1523]: will do!

**Comment by MartinuzziFrancesco** [comment]
I have updated the code and docs to address all the major comments

**Comment by Datseris** [comment]
Thanks a lot; there is a conflict preventing merging, can you please resolve it? I added another system and you just need to put it at the end or before your systems.

**Comment by Datseris** [comment]
fantastic, thanks for the contribution!

**Pull request merged**